File: src/pr_review_agent/analyzers/migration_analyzer.py
"""Migration analyzer — ported from migration-analyzer.ts.

Pure SQL parsing and risk assessment. Gets SQL from diff patches, not disk.
"""

from __future__ import annotations

import re

from pr_review_agent.models.migration import MigrationInfo, MigrationOperation
from pr_review_agent.models.pr import FileChange


def detect_migrations(files: list[FileChange]) -> list[MigrationInfo]:
    """Detect migration files in PR and analyze them."""
    migrations: list[MigrationInfo] = []

    for file in files:
        if not _is_migration_file(file.filename):
            continue

        sql = _extract_sql_from_patch(file.patch or "")
        if not sql:
            continue

        migration = analyze_migration(file.filename, sql)
        migrations.append(migration)

    return migrations


def _is_migration_file(filename: str) -> bool:
    """Check if a file is a migration file."""
    patterns = [
        r"prisma/migrations/.*/migration\.sql$",
        r"migrations/.*\.sql$",
        r"alembic/versions/.*\.py$",
        r"db/migrate/.*\.rb$",
    ]
    return any(re.search(p, filename) for p in patterns)


def _extract_sql_from_patch(patch: str) -> str:
    """Extract SQL content from a diff patch.

    Pulls only added lines (lines starting with +, excluding the +++ header).
    """
    if not patch:
        return ""

    lines: list[str] = []
    for line in patch.split("\n"):
        if line.startswith("+++"):
            continue
        if line.startswith("+"):
            lines.append(line[1:])  # Strip the leading +
        elif not line.startswith("-") and not line.startswith("@@"):
            # Context line — include as-is for parsing completeness
            lines.append(line)

    return "\n".join(lines)


def analyze_migration(path: str, sql: str) -> MigrationInfo:
    """Analyze a single migration file."""
    operations = parse_migration_sql(sql)
    risk_level = assess_migration_risk(operations)
    warnings = generate_migration_warnings(operations)
    rollback_complexity = assess_rollback_complexity(operations)

    # Extract migration name from path
    parts = path.split("/")
    name = parts[2] if len(parts) > 2 else "unknown"

    return MigrationInfo(
        path=path,
        name=name,
        sql=sql,
        risk_level=risk_level,
        operations=operations,
        warnings=warnings,
        rollback_complexity=rollback_complexity,
    )


def parse_migration_sql(sql: str) -> list[MigrationOperation]:
    """Parse SQL statements into migration operations."""
    operations: list[MigrationOperation] = []

    for line in sql.split("\n"):
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("--"):
            continue

        # CREATE TABLE
        if re.search(r"CREATE TABLE", trimmed, re.IGNORECASE):
            match = re.search(r'CREATE TABLE\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            operations.append(MigrationOperation(
                type="CREATE_TABLE", table=table, details=trimmed, destructive=False,
            ))

        # DROP TABLE
        elif re.search(r"DROP TABLE", trimmed, re.IGNORECASE):
            match = re.search(r'DROP TABLE\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            operations.append(MigrationOperation(
                type="DROP_TABLE", table=table, details=trimmed, destructive=True,
            ))

        # ALTER TABLE ADD COLUMN
        elif re.search(r"ALTER TABLE.*ADD COLUMN", trimmed, re.IGNORECASE):
            match = re.search(r'ALTER TABLE\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            not_null = "NOT NULL" in trimmed.upper()
            has_default = "DEFAULT" in trimmed.upper()
            operations.append(MigrationOperation(
                type="ADD_COLUMN",
                table=table,
                details=trimmed,
                destructive=not_null and not has_default,
            ))

        # ALTER TABLE DROP COLUMN
        elif re.search(r"ALTER TABLE.*DROP COLUMN", trimmed, re.IGNORECASE):
            match = re.search(r'ALTER TABLE\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            operations.append(MigrationOperation(
                type="DROP_COLUMN", table=table, details=trimmed, destructive=True,
            ))

        # ALTER TABLE ALTER COLUMN
        elif re.search(r"ALTER TABLE.*ALTER COLUMN", trimmed, re.IGNORECASE):
            match = re.search(r'ALTER TABLE\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            type_change = "TYPE" in trimmed.upper()
            operations.append(MigrationOperation(
                type="ALTER_COLUMN",
                table=table,
                details=trimmed,
                destructive=type_change,
            ))

        # CREATE INDEX
        elif re.search(r"CREATE.*INDEX", trimmed, re.IGNORECASE):
            match = re.search(r'ON\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            operations.append(MigrationOperation(
                type="CREATE_INDEX", table=table, details=trimmed, destructive=False,
            ))

        # ADD CONSTRAINT
        elif re.search(r"ADD CONSTRAINT", trimmed, re.IGNORECASE):
            match = re.search(r'ALTER TABLE\s+"?(\w+)"?', trimmed, re.IGNORECASE)
            table = match.group(1) if match else "unknown"
            operations.append(MigrationOperation(
                type="ADD_CONSTRAINT", table=table, details=trimmed, destructive=False,
            ))

    return operations


def assess_migration_risk(operations: list[MigrationOperation]) -> str:
    """Assess overall migration risk level."""
    has_drop_table = any(op.type == "DROP_TABLE" for op in operations)
    has_drop_column = any(op.type == "DROP_COLUMN" for op in operations)
    has_type_change = any(
        op.type == "ALTER_COLUMN" and "TYPE" in op.details.upper() for op in operations
    )

    if has_drop_table or has_drop_column or has_type_change:
        return "high"

    has_destructive = any(op.destructive for op in operations)
    has_add_not_null = any(
        op.type == "ADD_COLUMN"
        and "NOT NULL" in op.details.upper()
        and "DEFAULT" not in op.details.upper()
        for op in operations
    )

    if has_destructive or has_add_not_null:
        return "medium"

    return "low"


def generate_migration_warnings(operations: list[MigrationOperation]) -> list[str]:
    """Generate human-readable warnings for migration operations."""
    warnings: list[str] = []

    for op in operations:
        if op.type == "DROP_TABLE":
            warnings.append(f'DROP TABLE "{op.table}" will permanently delete all data')
        elif op.type == "DROP_COLUMN":
            warnings.append("DROP COLUMN will permanently delete data in that column")
        elif op.type == "ALTER_COLUMN" and "TYPE" in op.details.upper():
            warnings.append("Column type change may fail or lose data if incompatible")
        elif op.type == "ADD_COLUMN":
            if "NOT NULL" in op.details.upper() and "DEFAULT" not in op.details.upper():
                warnings.append("Adding NOT NULL column without DEFAULT may fail on existing data")
        elif op.type == "CREATE_INDEX":
            warnings.append("Creating index may take time on large tables")

    return warnings


def assess_rollback_complexity(operations: list[MigrationOperation]) -> str:
    """Assess how hard it would be to rollback this migration."""
    has_drop_table = any(op.type == "DROP_TABLE" for op in operations)
    has_drop_column = any(op.type == "DROP_COLUMN" for op in operations)

    if has_drop_table or has_drop_column:
        return "impossible"

    has_type_change = any(
        op.type == "ALTER_COLUMN" and "TYPE" in op.details.upper() for op in operations
    )
    if has_type_change:
        return "hard"

    has_constraints = any(op.type == "ADD_CONSTRAINT" for op in operations)
    if has_constraints:
        return "medium"

    return "easy"

File: context/pr-review-scripts-example/pr-review/migration-analyzer.ts
import { readFileSync, existsSync, readdirSync } from 'fs';
import { join } from 'path';
import type { MigrationInfo, MigrationOperation } from './types.js';

export function detectMigrations(files: any[]): MigrationInfo[] {
  const migrations: MigrationInfo[] = [];

  for (const file of files) {
    if (!file.filename.includes('prisma/migrations/')) continue;
    if (!file.filename.endsWith('migration.sql')) continue;

    try {
      const fullPath = join(process.cwd(), file.filename);
      if (existsSync(fullPath)) {
        const migration = analyzeMigration(file.filename, fullPath);
        migrations.push(migration);
      }
    } catch (err) {
      console.error(`Error analyzing migration ${file.filename}:`, err);
    }
  }

  // Also check for any new migration directories
  const migrationsDir = join(process.cwd(), 'prisma/migrations');
  if (existsSync(migrationsDir)) {
    const dirs = readdirSync(migrationsDir, { withFileTypes: true })
      .filter(d => d.isDirectory())
      .map(d => d.name);

    for (const dir of dirs) {
      const sqlFile = join(migrationsDir, dir, 'migration.sql');
      if (existsSync(sqlFile)) {
        const relativePath = `prisma/migrations/${dir}/migration.sql`;
        // Only add if not already in migrations array
        if (!migrations.some(m => m.path === relativePath)) {
          const isNew = files.some(f => f.filename.includes(dir));
          if (isNew) {
            migrations.push(analyzeMigration(relativePath, sqlFile));
          }
        }
      }
    }
  }

  return migrations;
}

function analyzeMigration(relativePath: string, fullPath: string): MigrationInfo {
  const sql = readFileSync(fullPath, 'utf-8');
  const operations = parseMigrationSQL(sql);
  const riskLevel = assessMigrationRisk(operations);
  const warnings = generateMigrationWarnings(operations);
  const rollbackComplexity = assessRollbackComplexity(operations);

  const name = relativePath.split('/')[2] || 'unknown';  // Extract migration name from path

  return {
    path: relativePath,
    name,
    sql,
    riskLevel,
    operations,
    warnings,
    rollbackComplexity
  };
}

function parseMigrationSQL(sql: string): MigrationOperation[] {
  const operations: MigrationOperation[] = [];
  const lines = sql.split('\n');

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('--')) continue;

    // CREATE TABLE
    if (trimmed.match(/CREATE TABLE/i)) {
      const tableMatch = trimmed.match(/CREATE TABLE\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      operations.push({
        type: 'CREATE_TABLE',
        table,
        details: trimmed,
        destructive: false
      });
    }

    // DROP TABLE
    else if (trimmed.match(/DROP TABLE/i)) {
      const tableMatch = trimmed.match(/DROP TABLE\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      operations.push({
        type: 'DROP_TABLE',
        table,
        details: trimmed,
        destructive: true
      });
    }

    // ALTER TABLE ADD COLUMN
    else if (trimmed.match(/ALTER TABLE.*ADD COLUMN/i)) {
      const tableMatch = trimmed.match(/ALTER TABLE\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      const notNull = trimmed.includes('NOT NULL');
      const hasDefault = trimmed.includes('DEFAULT');

      operations.push({
        type: 'ADD_COLUMN',
        table,
        details: trimmed,
        destructive: notNull && !hasDefault  // Dangerous if NOT NULL without DEFAULT
      });
    }

    // ALTER TABLE DROP COLUMN
    else if (trimmed.match(/ALTER TABLE.*DROP COLUMN/i)) {
      const tableMatch = trimmed.match(/ALTER TABLE\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      operations.push({
        type: 'DROP_COLUMN',
        table,
        details: trimmed,
        destructive: true
      });
    }

    // ALTER TABLE ALTER COLUMN
    else if (trimmed.match(/ALTER TABLE.*ALTER COLUMN/i)) {
      const tableMatch = trimmed.match(/ALTER TABLE\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      const typeChange = trimmed.includes('TYPE');

      operations.push({
        type: 'ALTER_COLUMN',
        table,
        details: trimmed,
        destructive: typeChange  // Type changes can be destructive
      });
    }

    // CREATE INDEX
    else if (trimmed.match(/CREATE.*INDEX/i)) {
      const tableMatch = trimmed.match(/ON\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      operations.push({
        type: 'CREATE_INDEX',
        table,
        details: trimmed,
        destructive: false
      });
    }

    // ADD CONSTRAINT
    else if (trimmed.match(/ADD CONSTRAINT/i)) {
      const tableMatch = trimmed.match(/ALTER TABLE\s+"?(\w+)"?/i);
      const table = tableMatch ? tableMatch[1] : 'unknown';
      operations.push({
        type: 'ADD_CONSTRAINT',
        table,
        details: trimmed,
        destructive: false  // Usually safe, but can fail if data doesn't meet constraint
      });
    }
  }

  return operations;
}

function assessMigrationRisk(operations: MigrationOperation[]): 'high' | 'medium' | 'low' {
  const hasDestructive = operations.some(op => op.destructive);
  const hasDropTable = operations.some(op => op.type === 'DROP_TABLE');
  const hasDropColumn = operations.some(op => op.type === 'DROP_COLUMN');
  const hasTypeChange = operations.some(op =>
    op.type === 'ALTER_COLUMN' && op.details.includes('TYPE')
  );

  if (hasDropTable || hasDropColumn || hasTypeChange) return 'high';
  if (hasDestructive) return 'medium';

  // Check for potentially risky operations
  const hasAddNotNull = operations.some(op =>
    op.type === 'ADD_COLUMN' && op.details.includes('NOT NULL') && !op.details.includes('DEFAULT')
  );

  if (hasAddNotNull) return 'medium';

  return 'low';
}

function generateMigrationWarnings(operations: MigrationOperation[]): string[] {
  const warnings: string[] = [];

  for (const op of operations) {
    switch (op.type) {
      case 'DROP_TABLE':
        warnings.push(`⚠️  DROP TABLE "${op.table}" will permanently delete all data`);
        break;

      case 'DROP_COLUMN':
        warnings.push(`⚠️  DROP COLUMN will permanently delete data in that column`);
        break;

      case 'ALTER_COLUMN':
        if (op.details.includes('TYPE')) {
          warnings.push(`⚠️  Column type change may fail or lose data if incompatible`);
        }
        break;

      case 'ADD_COLUMN':
        if (op.details.includes('NOT NULL') && !op.details.includes('DEFAULT')) {
          warnings.push(`⚠️  Adding NOT NULL column without DEFAULT may fail on existing data`);
        }
        break;

      case 'CREATE_INDEX':
        warnings.push(`ℹ️  Creating index may take time on large tables`);
        break;
    }
  }

  return warnings;
}

function assessRollbackComplexity(operations: MigrationOperation[]): 'easy' | 'medium' | 'hard' | 'impossible' {
  const hasDropTable = operations.some(op => op.type === 'DROP_TABLE');
  const hasDropColumn = operations.some(op => op.type === 'DROP_COLUMN');

  // Data deletion is impossible to rollback without backups
  if (hasDropTable || hasDropColumn) return 'impossible';

  const hasTypeChange = operations.some(op =>
    op.type === 'ALTER_COLUMN' && op.details.includes('TYPE')
  );

  // Type changes are hard to rollback
  if (hasTypeChange) return 'hard';

  const hasConstraints = operations.some(op => op.type === 'ADD_CONSTRAINT');
  if (hasConstraints) return 'medium';

  // Simple additions (tables, columns, indexes) are easy to rollback
  return 'easy';
}
